runtime/compare.c

```diff
@@ -107,20 +109,56 @@ static intnat compare_val(value v1, value v2, int total)
 #define GREATER 1
 #define UNORDERED ((intnat)1 << (8 * sizeof(value) - 1))
 
+static bool compare_poll_actions(struct compare_item** sp,
+                                 struct compare_stack* stk,
+                                 value *orig1, value *orig2,
+                                 value *v1, value *v2)
+{
+  value o1 = *orig1, o2 = *orig2;
+  if (caml_check_pending_actions()) {
+    compare_free_stack(stk);
```

Comment:
`caml_check_pending_actions()` can return false positives. It is considered that spurious polls are cheap, which simplifies various aspects of the implementation. One worry of backtracking pessimistically is that spurious polls are no longer necessarily cheap. It should be possible to avoid this with an internal `process_pending_actions` variant that returns whether something has actually been done; I will have to look further into this.

Reply:
That may complicate things too much, since we'd have to free the stack after a possible garbage collection.

Reply:
To be clearer, I have in mind doing the following: https://github.com/ocaml/ocaml/commit/7c104c555d47216ef285d9f10046c60f4dc07a98

However, the alternate path of scanning the compare stack would let one avoid doing this altogether.

Reply:
Ah, yes, it's not finished indeed. At least you get the idea. I do not have time to finish it right now but I can do it if needed.

Reply:
> The change to caml_do_pending_actions looks like it won't be necessary for the present PR

Review-wise, are we waiting to see where #12053 is headed, or is the plan to review and merge this one (with restarting) independently?